### PR TITLE
update(AppShell): Add breadcrumbs support to AppContext

### PR DESCRIPTION
**Pull request by milesj**
to: @milesj @stefhatcher @hayes @alecklandgraf

## Description

Adds breadcrumbs support to the `AppShell`. Also added a few new components:

- `Breadcrumbs` - Renders all crumbs in the context.
- `TrackBreadcrumb` - Adds and removes a crumb from the context.
- `PopToast` - Adds a toast to the context.

And also rewrote most of them using hooks.

## Motivation and Context

Ran into a use case where the breadcrumbs could be automated, so here we are.

## Testing

Aye.

## Screenshots

Screen Shot 2019-07-10 at 3 23 01 PM

## Checklist

- My code follows the style guide of this project.
- I have updated or added documentation accordingly.
- I have read the CONTRIBUTING document.

### CODE REVIEW

**Comment by milesj** [comment]
@stefhatcher This is good to go.

File: packages/app-shell/src/index.tsx [hunk 1]
```diff
@@ -35,10 +33,28 @@ export class AppShell extends React.Component<Props, State> {
...
     data: {},
     toasts: [],
   };
 
+  addBreadcrumb = (label: string, props: Breadcrumb['props'] = {}) => {
+    const id = uuid();
+
+    this.setState(prevState => ({
+      breadcrumbs: [
+        ...prevState.breadcrumbs.filter(crumb => crumb.id !== id),
```
Comment (stefhatcher) [inline packages/app-shell/src/index.tsx]: is this filter necessary? since `id` is created above for every `addBreadcrumb`?

Comment (stefhatcher) [reply packages/app-shell/src/index.tsx]: let's remove it and add it when/if we need it for add

File: packages/app-shell/src/components/Breadcrumbs/TrackBreadcrumb.tsx [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    return null;
+  }
+
+  useEffect(() => {
+    const id = context.addBreadcrumb(label, props);
+
+    return () => {
+      context.removeBreadcrumb(id);
+    };
+  }, [label]);
```
Comment (stefhatcher) [inline packages/app-shell/src/components/Breadcrumbs/TrackBreadcrumb.tsx]: do you also need to pass in `props` (based on that other lint error i saw yesterday)?

File: packages/app-shell/src/components/Breadcrumbs/index.tsx [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+import AppContext from '../AppContext';
+
+export { TrackBreadcrumb };
+
+export type Props = Omit<BaseProps, 'children'>;
+
+export default function Breadcrumbs(props: Props) {
+  const context = useContext(AppContext);
+
+  if (!context || context.breadcrumbs.length === 0) {
```
Comment (stefhatcher) [inline packages/app-shell/src/components/Breadcrumbs/index.tsx]: is `context.breadcrumbs` initial value `[]`?

File: packages/app-shell/src/components/Toasts/PopToast.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    if (props.danger || message instanceof Error) {
+      context.addFailureToast(message, props);
+    } else if (props.refresh) {
+      context.addRefreshToast(message, props);
+    } else if (props.success) {
+      context.addSuccessToast(message, props);
+    } else {
+      context.addInfoToast(message, props);
+    }
+  }, [message]);
```
Comment (stefhatcher) [inline packages/app-shell/src/components/Toasts/PopToast.tsx]: same q about adding `props` here for `props.danger`, `props.refresh`, and `props.success`

**Comment by milesj** [comment]
@stefhatcher Updated.

**Pull request merged**